nibe/connection/encoders.py

```diff
@@ -34,71 +35,141 @@
         """Encode coil data to bytes.
 
         :raises EncodeException: If encoding fails"""
         try:
             coil_data.validate()
-
-            return self._pad(self._get_parser(coil_data.coil), coil_data.raw_value)
+            return self.encode_raw_value(coil_data.coil.size, coil_data.raw_value)
         except (ValueError, ConstructError, ValidationError) as e:
             raise EncodeException(
                 f"Failed to encode {coil_data.coil.name} coil for value: {coil_data.value}, exception: {e}"
             )
 
-    def decode(self, coil: Coil, raw: bytes) -> CoilData:
+    @abstractmethod
+    def decode_raw_value(self, size: str, raw: _RawDataT) -> Union[int, None]:
+        pass
+
+    def decode(self, coil: Coil, raw: _RawDataT) -> CoilData:
         """Decode coil data from bytes.
 
         :raises DecodeException: If decoding fails"""
         try:
-            parser = self._get_parser(coil)
-            assert parser.sizeof() <= len(
-                raw
-            ), f"Invalid raw data size: given {len(raw)}, expected at least {parser.sizeof()}"
-            value = parser.parse(raw)
-            if self._is_hitting_integer_limit(coil, value):
-                return CoilData(coil, None)
-
-            return CoilData.from_raw_value(coil, value)
+            return CoilData.from_raw_value(coil, self.decode_raw_value(coil.size, raw))
         except (ValueError, AssertionError, ConstructError, ValidationError) as e:
             raise DecodeException(
                 f"Failed to decode {coil.name} coil from raw: {hexlify(raw).decode('utf-8')}, exception: {e}"
             ) from e
 
-    def _is_hitting_integer_limit(self, coil: Coil, int_value: int):
-        if coil.size == "u8" and int_value == 0xFF:
-            return True
-        if coil.size == "s8" and int_value == -0x80:
-            return True
-        if coil.size == "u16" and int_value == 0xFFFF:
-            return True
-        if coil.size == "s16" and int_value == -0x8000:
-            return True
-        if coil.size == "u32" and int_value == 0xFFFFFFFF:
-            return True
-        if coil.size == "s32" and int_value == -0x80000000:
-            return True
-
-        return False
-
-    def _get_parser(self, coil: Coil) -> Construct:
-        if coil.size in ["u32", "s32"] and self.word_swap is None:
+    def _is_hitting_integer_limit(self, size: str, int_value: int):
+        limit = integer_limit[size]
+        if limit < 0:
+            return int_value <= limit
+        return int_value >= limit
+
+
+class CoilDataEncoderNibeGw(CoilDataEncoder[bytes]):
+    """Encode and decode coil data."""
+
+    word_swap: Optional[bool] = None
+
+    def __init__(self, word_swap: Optional[bool] = None):
+        self.word_swap = word_swap
+
+    def encode_raw_value(self, size: str, raw_value: Union[int, None]) -> bytes:
+        """Encode coil data to bytes."""
+        if raw_value is None:
+            raw_value = integer_limit[size]
```

Comment:
Why you want to encode an invalid value?

Reply:
I prefer if we can roundtrip test data == decode(encode(data)). Since we consider integer limit values as signifying invalid/none on decode, i want to be able to encode same thing.

Reply:
That said. That check should likely move up one level.